extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js

```diff
@@ -343,20 +343,23 @@ function performServiceWorkerOptimizations(registration, win, element) {
  * Whenever a new service worker is activated, controlled page will send
  * the used AMP scripts and the self's URL to service worker to be cached.
  * @param {!Window} win
+ * @param {ServiceWorkerRegistration} registration
  */
-function sendAmpScriptToSwOnFirstVisit(win) {
+function sendAmpScriptToSwOnFirstVisit(win, registration) {
   if ('performance' in win) {
     // Fetch all AMP-scripts used on the page
     const ampScriptsUsed = win.performance.getEntriesByType('resource')
         .filter(item => item.initiatorType === 'script' &&
           startsWith(item.name, urls.cdn))
         .map(script => script.name);
-    const controllerSw = win.navigator.serviceWorker.controller;
+    const activeSW = registration.active;
     // using convention from https://github.com/redux-utilities/flux-standard-action.
-    controllerSw.postMessage(JSON.stringify(dict({
-      'type': 'AMP__FIRST-VISIT-CACHING',
-      'payload': ampScriptsUsed,
-    })));
+    if (activeSW.postMessage) {
```

Comment:
Why check this? I feel like `activeSw` is more likely to be nullish than `postMessage`

Reply:
well actually activeSw's presence is guarded by this: https://github.com/ampproject/amphtml/blob/master/extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js#L310-L320

And i guess i was just being defensive by checking postMessage's presence. I think there were some initial browsers which did not supported postMessaging to SW (chrome 40-44).